### PR TITLE
Ataymano/logger1

### PR DESCRIPTION
**Pull request by ataymano**
cb logger.

### CODE REVIEW

File: bindings/cs/rl.net.native/rl.net.callback.h [hunk 1]
```diff
@@ -0,0 +1,12 @@
+#pragma once
+
+#include "rl.net.native.h"
+
+namespace rl_net_native {
+  namespace constants {
+    const char *const BINDING_TRACE_LOGGER = "BINDING_TRACE_LOGGER";
+  }
+  typedef void(*background_error_callback_t)(const reinforcement_learning::api_status&);
```
Comment (jackgerrits) [inline bindings/cs/rl.net.native/rl.net.callback.h]: Nit: prefer using

File: bindings/cs/rl.net/DecisionResponse.cs [hunk 2]
```diff
@@ -98,6 +106,19 @@ public DecisionResponse() : base(new New<DecisionResponse>(NativeMethods.CreateD
...
+        {
+            var result = new DecisionResponse();
+            DecisionResponseSetModelId(result.DangerousGetHandle(), modelId);
+            return result;
+        }
+
+
+        public DecisionResponse AddSlot(string eventId, IList<int> actions, IList<float> probabilities)
+        {
+            return this;
```
Comment (kumpera) [inline bindings/cs/rl.net/DecisionResponse.cs]: This looks wrong.

File: bindings/cs/rl.net/RlLogger.cs [hunk 3]
```diff
@@ -0,0 +1,337 @@
...
+
+            [DllImport("rl.net.native.dll", EntryPoint = "RlLoggerLogCbInteraction")]
+            private static extern int RlLoggerLogCbInteractionNative(IntPtr logger, IntPtr contextJson, IntPtr rankingResponse, IntPtr apiStatus);
+
+            internal static Func<IntPtr, IntPtr, IntPtr, IntPtr, int> RlLoggerLogCbInteractionOverride { get; set; }
+
+            [DllImport("rl.net.native.dll", EntryPoint = "RlLoggerLogCcbInteraction")]
+            private static extern int RlLoggerLogCcbInteractionNative(IntPtr logger, IntPtr contextJson, IntPtr decisionResponse, IntPtr apiStatus);
+
+            internal static Func<IntPtr, IntPtr, IntPtr, IntPtr, int> RlLoggerLogCcbInteractionOverride { get; set; }
```
Comment (kumpera) [inline bindings/cs/rl.net/RlLogger.cs]: This is not used anywhere

File: bindings/cs/rl.net/RlLogger.cs [hunk 4]
```diff
@@ -0,0 +1,337 @@
...
+            {
+                return NativeMethods.RlLoggerLogOutcomeJson(liveModel, new IntPtr(eventIdUtf8Bytes), new IntPtr(outcomeJsonUtf8Bytes), apiStatus);
+            }
+        }
+
+        private void WrapStatusAndRaiseBackgroundError(IntPtr apiStatusHandle)
+        {
+            using (ApiStatus status = new ApiStatus(apiStatusHandle))
+            {
+                EventHandler<ApiStatus> trargetEventLocal = this.BackgroundErrorInternal;
```
Comment (kumpera) [inline bindings/cs/rl.net/RlLogger.cs]: typo.

Comment (kumpera) [inline bindings/cs/rl.net/RlLogger.cs]:
```suggestion
                EventHandler<ApiStatus> targetEventLocal = this.BackgroundErrorInternal;
```

File: bindings/cs/rl.net/RlLogger.cs [hunk 5]
```diff
@@ -0,0 +1,337 @@
...
+            using (ApiStatus apiStatus = new ApiStatus())
+            {
+                if (!this.TryLog(eventId, outcome, apiStatus))
+                {
+                    throw new RLException(apiStatus);
+                }
+            }
+        }
+
+        private event EventHandler<ApiStatus> BackgroundErrorInternal;
```
Comment (kumpera) [inline bindings/cs/rl.net/RlLogger.cs]: Why does this exists?

File: bindings/cs/rl.net/RlLogger.cs [hunk 6]
```diff
@@ -0,0 +1,337 @@
...
+            {
+                this.BackgroundErrorInternal += value;
+            }
+            remove
+            {
+                this.BackgroundErrorInternal -= value;
+            }
+        }
+
+        private event EventHandler<TraceLogEventArgs> OnTraceLoggerEventInternal;
```
Comment (kumpera) [inline bindings/cs/rl.net/RlLogger.cs]: Why does this exists?

File: bindings/cs/rl.net/RlLogger.cs [hunk 7]
```diff
@@ -0,0 +1,337 @@
...
+
+        // TODO:
+        /// <remarks>
+        /// Add/remove here is not thread safe.
+        /// </remarks>
+        public event EventHandler<TraceLogEventArgs> TraceLoggerEvent
```
Comment (kumpera) [inline bindings/cs/rl.net/RlLogger.cs]: Make this thread safe by using a sync block.

File: bindings/cs/rl.net/RlLogger.cs (hunk 5 above)
Comment (kumpera) [reply bindings/cs/rl.net/RlLogger.cs]: Delete it and let `BackgroundError` use the default implementation.

File: bindings/cs/rl.net/RlLogger.cs (hunk 6 above)
Comment (kumpera) [reply bindings/cs/rl.net/RlLogger.cs]: There is no reason for this to be an event. Make it a delegate.

File: bindings/cs/rl.net/RlLogger.cs [hunk 8]
```diff
@@ -0,0 +1,337 @@
...
+                if (!this.TryLog(eventId, outcome, apiStatus))
+                {
+                    throw new RLException(apiStatus);
+                }
+            }
+        }
+
+        public bool Log(string eventId, string outcome, ApiStatus apiStatus = null)
+        {
+            return this.TryLog(eventId, outcome, apiStatus);
```
Comment (kumpera) [inline bindings/cs/rl.net/RlLogger.cs]: The non try version of Log should throw on failure and return void.

File: include/rl_logger.h [hunk 9]
```diff
@@ -0,0 +1,181 @@
...
+   * @brief Interface class for the logging API.
+   *
+   * - (1) Instantiate and Initialize
+   * - (2) log() to log contextual bandints evend ([event_id], context, response, outcome)
+   */
+  class rl_logger {
+  public:
+    /**
+     * @brief Error callback function.
+     * When live_model is constructed, a background error callback and a
```
Comment (rajan-chari) [inline include/rl_logger.h]: >live_model [](start = 12, length = 10)

rl_logger

File: include/rl_logger.h [hunk 10]
```diff
@@ -0,0 +1,181 @@
...
+     * context (void*) is registered. If there is an error in the background thread,
+     * error callback will get invoked with api_status and the context (void*).
+     *
+     * NOTE: Error callback will get invoked in a background thread.
+     */
+    using error_fn = void(*)(const api_status&, void*);
+
+    /**
+     * @brief Construct a new live model object.
```
Comment (rajan-chari) [inline include/rl_logger.h]: >live model [](start = 30, length = 10)

logger

File: include/rl_logger.h [hunk 11]
```diff
@@ -0,0 +1,181 @@
...
+     *
+     * @param config Name-Value based configuration
+     * @param fn Error callback for handling errors in background thread
+     * @param err_context Context passed back during Error callback
+     * @param t_factory Transport factory.  The default transport factory is initialized with a
+     *                  REST based transport that gets data from an Azure storage account
+     * @param m_factory Model factory.  The default model factory hydrates vw models
```
Comment (rajan-chari) [inline include/rl_logger.h]: >m_factory [](start = 14, length = 9)

No model factory.

File: include/rl_logger.h [hunk 12]
```diff
@@ -0,0 +1,181 @@
...
+     *                    used for local inference.
+     * @param sender_factory Sender factory.  The default factory provides two senders, one for
+     *                       interaction and the other for observation which logs to Event Hub.
+     */
```
Comment (rajan-chari) [inline include/rl_logger.h]: Needs time_provider_factory.  Probably missing from live_model also.

File: include/rl_logger.h [hunk 13]
```diff
@@ -0,0 +1,181 @@
...
+    rl_logger(
+      const utility::configuration& config,
+      error_fn fn = nullptr,
+      void* err_context = nullptr,
+      trace_logger_factory_t* trace_factory = &trace_logger_factory,
+      sender_factory_t* s_factory = &sender_factory,
+      time_provider_factory_t* time_prov_factory = &time_provider_factory);
+
+    /**
+     * @brief Initialize inference library.
```
Comment (rajan-chari) [inline include/rl_logger.h]: >inference [](start = 25, length = 9)

logger

File: include/rl_logger.h [hunk 14]
```diff
@@ -0,0 +1,181 @@
...
+     * Initialize the library and start the background threads used for
+     * model managment and sending actions and outcomes to the online trainer
```
Comment (rajan-chari) [inline include/rl_logger.h]: >model managment and [](start = 7, length = 19)

No model management

File: include/rl_logger.h [hunk 15]
```diff
@@ -0,0 +1,181 @@
...
+     * @param status  Optional field with detailed string description if there is an error
+     * @return int Return error code.  This will also be returned in the api_status object
+     */
+    int log(const char * event_id, float outcome, api_status* status = nullptr);
+
+    template<typename ErrCntxt>
+    using error_fn_t = void(*)(const api_status&, ErrCntxt*);
+
+    /**
+     * @brief Construct a new live model object.
```
Comment (rajan-chari) [inline include/rl_logger.h]: >live model [](start = 30, length = 10)

logger

File: include/rl_logger.h [hunk 16]
```diff
@@ -0,0 +1,181 @@
...
+     *
+     * @tparam ErrCntxt Context type used in error callback.
+     * @param config Name-Value based configuration
+     * @param fn Error callback for handling errors in background thread
+     * @param err_context Context passed back during Error callback
+     * @param t_factory Transport factory.  The default transport factory is initialized with a
+     *                  REST based transport that gets data from an Azure storage account
+     * @param m_factory Model factory.  The default model factory hydrates vw models
```
Comment (rajan-chari) [inline include/rl_logger.h]: > m_factory [](start = 13, length = 10)

No model factory

Comment (rajan-chari) [inline include/rl_logger.h]: > m_factory [](start = 13, length = 10)

time_provider_factory needed here and in live_model.h

File: include/rl_logger.h [hunk 17]
```diff
@@ -0,0 +1,181 @@
...
+    rl_logger(
+      const utility::configuration& config,
+      error_fn_t<ErrCntxt> fn = nullptr,
+      ErrCntxt* err_context = nullptr,
+      trace_logger_factory_t* trace_factory = &trace_logger_factory,
+      sender_factory_t* s_factory = &sender_factory,
+      time_provider_factory_t* time_prov_factory = &time_provider_factory);
+
+    /**
+     * @brief Move constructor for live model object.
```
Comment (rajan-chari) [inline include/rl_logger.h]: >live model [](start = 35, length = 10)

logger

File: include/rl_logger.h [hunk 18]
```diff
@@ -0,0 +1,181 @@
...
+     */
+    rl_logger& operator=(rl_logger&& other);
+   
+    rl_logger(const rl_logger&) = delete;       //! Prevent accidental copy, since destructor will deallocate the implementation
+    rl_logger& operator=(rl_logger&) = delete;  //! Prevent accidental copy, since destructor will deallocate the implementation
+
+    ~rl_logger();
+  private:
+    std::unique_ptr<rl_logger_impl> _pimpl;  //! The actual implementation details are forwarded to this object (PIMPL pattern)
+    bool _initialized = false;                //! Guard to ensure that live_model is properly initialized. i.e. init() was called and successfully initialized.
```
Comment (rajan-chari) [inline include/rl_logger.h]: >live_model [](start = 71, length = 10)

logger

File: include/rl_logger.h [hunk 19]
```diff
@@ -0,0 +1,181 @@
...
+  };
+
+  /**
+   * @brief Construct a new live model object.
```
Comment (rajan-chari) [inline include/rl_logger.h]: >live model [](start = 28, length = 10)

logger

File: include/rl_logger.h [hunk 20]
```diff
@@ -0,0 +1,181 @@
...
+   *
+   * @tparam ErrCntxt Context type used in error callback.
+   * @param config Name-Value based configuration
+   * @param fn Error callback for handling errors in background thread
+   * @param err_context Context passed back during Error callback
+   * @param t_factory Transport factory.  The default transport factory is initialized with a
+   *                  REST based transport that gets data from an Azure storage account
+   * @param m_factory Model factory.  The default model factory hydrates vw models
```
Comment (rajan-chari) [inline include/rl_logger.h]: >m_factory [](start = 12, length = 9)

No model factory.  Need time_provider_factory.

File: bindings/cs/rl.net/RlLogger.cs (hunk 6 above)
Comment (lokitoth) [reply bindings/cs/rl.net/RlLogger.cs]: I am not sure I follow what you mean by changing this to a delegate? This is used to enable native=>managed transition hooking/unhooking inside of the add/remove handlers for TraceLoggerEvent.

File: rlclientlib/rl_logger_impl.h [hunk 21]
```diff
@@ -0,0 +1,99 @@
+#pragma once
+#include "logger/event_logger.h"
+#include "model_mgmt.h"
```
Comment (rajan-chari) [inline rlclientlib/rl_logger_impl.h]: Do we need model_mgmt for logging only?

File: rlclientlib/rl_logger_impl.cc [hunk 22]
```diff
@@ -0,0 +1,204 @@
...
+#include "utility/context_helper.h"
+#include "sender.h"
+#include "api_status.h"
+#include "configuration.h"
+#include "error_callback_fn.h"
+#include "ranking_response.h"
+
+#include "err_constants.h"
+#include "constants.h"
+#include "vw_model/safe_vw.h"
```
Comment (rajan-chari) [inline rlclientlib/rl_logger_impl.cc]: >vw_model [](start = 10, length = 8)

not needed?

File: rlclientlib/rl_logger_impl.cc [hunk 23]
```diff
@@ -0,0 +1,204 @@
...
+#include "trace_logger.h"
+#include "explore_internal.h"
```
Comment (rajan-chari) [inline rlclientlib/rl_logger_impl.cc]: >explore_internal [](start = 10, length = 16)

not needed?

File: rlclientlib/rl_logger_impl.cc [hunk 24]
```diff
@@ -0,0 +1,204 @@
...
+#include "hash.h"
+#include "factory_resolver.h"
+#include "logger/preamble_sender.h"
+#include "sampling.h"
```
Comment (rajan-chari) [inline rlclientlib/rl_logger_impl.cc]: >sampling [](start = 10, length = 8)

not needed?

File: rlclientlib/rl_logger_impl.cc [hunk 25]
```diff
@@ -0,0 +1,204 @@
...
+#include <boost/uuid/uuid_io.hpp>
+#include <boost/uuid/random_generator.hpp>
+
+// Some namespace changes for more concise code
+namespace e = exploration;
+using namespace std;
+
+namespace reinforcement_learning {
+  // Some namespace changes for more concise code
+  namespace m = model_management;
```
Comment (rajan-chari) [inline rlclientlib/rl_logger_impl.cc]: >model_management [](start = 16, length = 16)

not needed?

File: bindings/cs/rl.net/DecisionResponse.cs (hunk 2 above)
Comment (ataymano) [reply bindings/cs/rl.net/DecisionResponse.cs]: fixed

File: bindings/cs/rl.net/RlLogger.cs (hunk 7 above)
Comment (ataymano) [reply bindings/cs/rl.net/RlLogger.cs]: Switched to RlLoggerThreadUnsafe/ThreadLogger implementations exactly as we have for LiveModel. If there are remaining issues that has to be fixed in both places, let's do it in separate PR.

File: bindings/cs/rl.net/RlLogger.cs (hunk 6 above)
Comment (ataymano) [reply bindings/cs/rl.net/RlLogger.cs]: Right now this pattern exactly replicating the one from LiveModel. If there are remaining issues that has to be fixed in both places, let's do it in separate PR.

File: bindings/cs/rl.net/RlLogger.cs (hunk 5 above)
Comment (ataymano) [reply bindings/cs/rl.net/RlLogger.cs]: Right now this pattern exactly replicating the one from LiveModel. If there are remaining issues that has to be fixed in both places, let's do it in separate PR.

**Pull request closed without merging**